File: src/core/invisible.ts
import type { Challenge, ImRobotToken } from './types'
import { solveChallenge } from './solver'
import { createToken } from './challenge'

export interface InvisibleVerifyOptions {
  /** URL of the challenge endpoint (e.g., 'https://api.example.com/imrobot/challenge') */
  challengeUrl: string
  /** URL of the verify endpoint (e.g., 'https://api.example.com/imrobot/verify') */
  verifyUrl: string
  /** Agent identifier */
  agentId?: string
  /** Number of retry attempts (default: 3) */
  maxRetries?: number
  /** Timeout per request in ms (default: 10000) */
  timeout?: number
}

export interface InvisibleVerifyResult {
  /** Whether verification succeeded */
  success: boolean
  /** The proof token (if successful) */
  proofToken?: string
  /** The imrobot token (if successful) */
  token?: ImRobotToken
  /** Error message (if failed) */
  error?: string
  /** Number of attempts made */
  attempts: number
  /** Time taken to solve the challenge in ms */
  solveTime?: number
  /** Total time taken in ms */
  totalTime: number
}

/**
 * Invisible (zero-UI) agent verification.
 *
 * Fetches a challenge from the server, solves it locally, submits the answer,
 * and returns a proof token. No user interaction required.
 *
 * @example
 * ```typescript
 * import { invisibleVerify } from 'imrobot/core'
 *
 * const result = await invisibleVerify({
 *   challengeUrl: 'https://api.example.com/imrobot/challenge',
 *   verifyUrl: 'https://api.example.com/imrobot/verify',
 * })
 *
 * if (result.success) {
 *   // Use result.proofToken in X-Agent-Proof header
 *   fetch('/api/data', {
 *     headers: { 'X-Agent-Proof': result.proofToken! }
 *   })
 * }
 * ```
 */
export async function invisibleVerify(
  options: InvisibleVerifyOptions,
): Promise<InvisibleVerifyResult> {
  const maxRetries = options.maxRetries ?? 3
  const timeout = options.timeout ?? 10_000
  const startTime = Date.now()
  let attempts = 0

  for (let attempt = 0; attempt < maxRetries; attempt++) {
    attempts++
    try {
      // 1. Fetch challenge
      const challengeResponse = await fetchWithTimeout(
        options.challengeUrl,
        {
          method: 'GET',
          headers: { Accept: 'application/json' },
        },
        timeout,
      )

      if (!challengeResponse.ok) {
        throw new Error(`Challenge request failed: ${challengeResponse.status}`)
      }

      const challenge: Challenge = await challengeResponse.json()

      // 2. Solve locally
      const solveStart = Date.now()
      const answer = solveChallenge(challenge)
      const solveTime = Date.now() - solveStart

      // 3. Submit answer
      const verifyResponse = await fetchWithTimeout(
        options.verifyUrl,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            challenge,
            answer,
            agentId: options.agentId,
          }),
        },
        timeout,
      )

      if (!verifyResponse.ok) {
        const errorBody = await verifyResponse.json().catch(() => ({}))
        throw new Error(
          `Verify failed: ${(errorBody as Record<string, string>).reason ?? verifyResponse.status}`,
        )
      }

      const result = (await verifyResponse.json()) as {
        valid: boolean
        proofToken?: string
        elapsed?: number
        suspicious?: boolean
      }

      if (result.valid) {
        const token = createToken(challenge, answer, solveStart)
        return {
          success: true,
          proofToken: result.proofToken,
          token,
          attempts,
          solveTime,
          totalTime: Date.now() - startTime,
        }
      }

      // If not valid, retry
    } catch (error) {
      // Retry on network errors
      if (attempt === maxRetries - 1) {
        return {
          success: false,
          error: error instanceof Error ? error.message : 'Unknown error',
          attempts,
          totalTime: Date.now() - startTime,
        }
      }
      // Exponential backoff: 100ms, 200ms, 400ms...
      await sleep(100 * Math.pow(2, attempt))
    }
  }

  return {
    success: false,
    error: 'Max retries exceeded',
    attempts,
    totalTime: Date.now() - startTime,
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

async function fetchWithTimeout(
  url: string,
  init: RequestInit,
  timeout: number,
): Promise<Response> {
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), timeout)
  try {
    return await fetch(url, { ...init, signal: controller.signal })
  } finally {
    clearTimeout(timer)
  }
}

File: python/imrobot/invisible.py
"""
Invisible (zero-UI) verification for AI agents.

Port of src/core/invisible.ts -- agents verify themselves
programmatically without any UI interaction.
"""

from __future__ import annotations

import json
import time
import urllib.request
import urllib.error
from dataclasses import dataclass
from typing import Optional

from .core import solve_challenge
from .types import Challenge


@dataclass
class InvisibleVerifyResult:
    """Result of invisible verification."""

    success: bool
    proof_token: Optional[str] = None
    error: Optional[str] = None
    attempts: int = 0
    elapsed_ms: int = 0


def invisible_verify(
    challenge_url: str,
    verify_url: str,
    agent_id: str = "python-agent",
    max_retries: int = 3,
    timeout: int = 10,
) -> InvisibleVerifyResult:
    """
    Perform invisible (zero-UI) agent verification.

    Fetches a challenge from the server, solves it locally, and submits
    the answer for verification. Returns a proof token on success.

    Args:
        challenge_url: URL to GET a challenge (e.g., "https://api.example.com/imrobot/challenge").
        verify_url: URL to POST the answer (e.g., "https://api.example.com/imrobot/verify").
        agent_id: Identifier for this agent.
        max_retries: Maximum number of retry attempts.
        timeout: HTTP request timeout in seconds.

    Returns:
        InvisibleVerifyResult with success status and proof token.

    Example::

        result = invisible_verify(
            challenge_url="https://api.example.com/imrobot/challenge",
            verify_url="https://api.example.com/imrobot/verify",
            agent_id="my-bot-v1",
        )
        if result.success:
            headers = {"X-Agent-Proof": result.proof_token}
    """
    start = int(time.time() * 1000)

    for attempt in range(1, max_retries + 1):
        try:
            # 1. Fetch challenge
            req = urllib.request.Request(
                challenge_url,
                headers={"Accept": "application/json"},
            )
            with urllib.request.urlopen(req, timeout=timeout) as resp:
                challenge_data = json.loads(resp.read().decode("utf-8"))

            challenge = Challenge.from_dict(challenge_data)

            # 2. Solve
            answer = solve_challenge(challenge)

            # 3. Submit answer
            body = json.dumps({
                "challenge": challenge_data,
                "answer": answer,
                "agentId": agent_id,
            }).encode("utf-8")

            verify_req = urllib.request.Request(
                verify_url,
                data=body,
                headers={
                    "Content-Type": "application/json",
                    "Accept": "application/json",
                },
                method="POST",
            )

            with urllib.request.urlopen(verify_req, timeout=timeout) as resp:
                result = json.loads(resp.read().decode("utf-8"))

            if result.get("valid"):
                elapsed = int(time.time() * 1000) - start
                return InvisibleVerifyResult(
                    success=True,
                    proof_token=result.get("proofToken"),
                    attempts=attempt,
                    elapsed_ms=elapsed,
                )

        except (urllib.error.URLError, json.JSONDecodeError, Exception) as e:
            if attempt == max_retries:
                elapsed = int(time.time() * 1000) - start
                return InvisibleVerifyResult(
                    success=False,
                    error=str(e),
                    attempts=attempt,
                    elapsed_ms=elapsed,
                )

    elapsed = int(time.time() * 1000) - start
    return InvisibleVerifyResult(
        success=False,
        error="Max retries exceeded",
        attempts=max_retries,
        elapsed_ms=elapsed,
    )
